From 2042668bb62ede85f5f0c82b4d0ce8117097fcb5 Mon Sep 17 00:00:00 2001
From: Stefan Chulski <stefanc@marvell.com>
Date: Wed, 28 Jun 2017 14:36:10 +0300
Subject: [PATCH 1073/1345] fix: net: mvpp2x: swap dropped packets RX desc if
 BIG_ENDIAN set

commit  5f6093bde1c97086fdb9183c2519ec7cdf88c073 from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

In BIG_ENDIAN mode all descriptors should be swapped.
Swap RX descriptor in RX drop procedure, under BIG_ENDIAN ifdef
(similar to regular RX path).
Issue could cause BM refill with wrong buffer address and memory
overrun.

Change-Id: I67ad9ef0495e7d8c950d8af550614bcb0c010fcd
Signed-off-by: Stefan Chulski <stefanc@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/40929
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Reviewed-by: Yuval Caduri <cyuval@marvell.com>
Reviewed-by: Omri Itach <omrii@marvell.com>
Reviewed-by: Hanna Hawa <hannah@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_main.c |    8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_main.c b/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_main.c
index 4d5e24b..01b7f26b4 100644
--- a/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_main.c
+++ b/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_main.c
@@ -1293,6 +1293,14 @@ static void mv_pp2x_rxq_drop_pkts(struct mv_pp2x_port *port,
 	for (i = 0; i < rx_received; i++) {
 		struct mv_pp2x_rx_desc *rx_desc =
 			mv_pp2x_rxq_next_desc_get(rxq);
+
+#if defined(__BIG_ENDIAN)
+		if (port->priv->pp2_version == PPV21)
+			mv_pp21_rx_desc_swap(rx_desc);
+		else
+			mv_pp22_rx_desc_swap(rx_desc);
+#endif /* __BIG_ENDIAN */
+
 		bm_pool = &port->priv->bm_pools[MVPP2_RX_DESC_POOL(rx_desc)];
 
 		if (port->priv->pp2_version == PPV21)
-- 
1.7.9.5

